From 40cbbb1fdd087f2f02b617bc164bd238fd7ba595 Mon Sep 17 00:00:00 2001
From: Nam Ninh <nam.ninh@windriver.com>
Date: Thu, 26 Feb 2015 14:37:19 -0500
Subject: [PATCH 18/18] qemu: update msi affinity in pci-assign passthrough
 device

Issue: OVP-2305

When intlat test is run in OVP affined guest, the IRQs are
affined to vCPU1 on the guest which is pinned to physical
core 2, so all Altera FPGA IRQs should be on core 2 on the
host. However, we observed the IRQs are moved to another
core dirrent than core 2 when the test is started. The problem
is in qemu where MSI affinity is not updated to the right core
in assigned_dev_update_msi_msg.

Signed-off-by: Nam Ninh <nam.ninh@windriver.com>
[ywei: rebase to qemu 2.4 ]
Signed-off-by: Yunguo Wei <yunguo.wei@windriver.com>
[Adjust code context
 Add pci_dev as parameter to kvm_irqchip_update_msi_route according to upstream]
Upstream-Status: Pending
Signed-off-by: He Zhe <zhe.he@windriver.com>
---
 hw/i386/kvm/pci-assign.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/hw/i386/kvm/pci-assign.c b/hw/i386/kvm/pci-assign.c
index 24865f50..39a7d758 100644
--- a/hw/i386/kvm/pci-assign.c
+++ b/hw/i386/kvm/pci-assign.c
@@ -1018,13 +1018,19 @@ static void assigned_dev_update_msi_msg(PCIDevice *pci_dev)
     uint8_t ctrl_byte = pci_get_byte(pci_dev->config + pci_dev->msi_cap +
                                      PCI_MSI_FLAGS);
 
+    MSIMessage msg;
+    uint32_t dest;
+
     if (assigned_dev->assigned_irq_type != ASSIGNED_IRQ_MSI ||
         !(ctrl_byte & PCI_MSI_FLAGS_ENABLE)) {
         return;
     }
 
-    kvm_irqchip_update_msi_route(kvm_state, assigned_dev->msi_virq[0],
-                                 msi_get_message(pci_dev, 0), pci_dev);
+    msg = msi_get_message(pci_dev, 0);
+    kvm_irqchip_update_msi_route(kvm_state, assigned_dev->msi_virq[0], msg, pci_dev);
+    dest = ((uint32_t)msg.address & MSI_ADDR_DEST_ID_MASK) >> MSI_ADDR_DEST_ID_SHIFT;
+    kvm_msix_update_affinity(assigned_dev, 0, dest);
+
     kvm_irqchip_commit_routes(kvm_state);
 }
 
-- 
2.11.0

